### PR TITLE
feat: Add Currency Pair Providers and Deviations parameters

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

closes: #241 

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/ojo/oracle/v1/genesis.proto [hunk 1]
```diff
@@ -25,7 +25,7 @@ message GenesisState {
       [ (gogoproto.nullable) = false ];
   repeated PriceStamp medians = 7 [ (gogoproto.nullable) = false ];
   repeated PriceStamp historic_prices = 8 [ (gogoproto.nullable) = false ];
-  repeated PriceStamp medianDeviations = 9 [ (gogoproto.nullable) = false ];
+  repeated PriceStamp median_deviations = 9 [ (gogoproto.nullable) = false ];
```
Comment (adamewozniak) [inline proto/ojo/oracle/v1/genesis.proto:28]: nice. does this change the API for @dixitaniket 's relayer?

Comment (rbajollari) [reply proto/ojo/oracle/v1/genesis.proto:28]: I don't think it should but I'd like him to confirm just to make sure

File: x/oracle/types/currency_pair_providers.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	"gopkg.in/yaml.v3"
+)
+
+// String implements fmt.Stringer interface
+func (cpp CurrencyPairProviders) String() string {
+	out, _ := yaml.Marshal(cpp)
+	return string(out)
+}
+
+func (cpp CurrencyPairProviders) Equal(cpp2 *CurrencyPairProviders) bool {
```
Comment (zarazan) [inline x/oracle/types/currency_pair_providers.go:16]: Does this need to check the `PairAddressProvider` for equality as well?

**Review by zarazan — APPROVED** [review]
Looks great, only one comment. My only other comment is that the params are getting very large and are kind of a pain to update since you have to provide empty values for all the values you don't want to update. Could be valuable to eventually make specific transactions for updating it.

**Comment by adamewozniak** [comment]
@zarazan can I get another review here

**Review by zarazan — APPROVED** [review]
Looks great. Like I mentioned before our params are getting pretty huge and I think it would be good for us to move off of using `cosmos-sdk/x/params` all together but fun for another time.

File: x/oracle/types/currency_deviation_thresholds.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	out, _ := yaml.Marshal(cdt)
+	return string(out)
+}
+
+func (cdt CurrencyDeviationThreshold) Equal(cdt2 *CurrencyDeviationThreshold) bool {
+	if !strings.EqualFold(cdt.BaseDenom, cdt2.BaseDenom) || !strings.EqualFold(cdt.Threshold, cdt2.Threshold) {
+		return false
+	}
+
+	return true
```
Comment (zarazan) [inline x/oracle/types/currency_deviation_thresholds.go:20]: nit: could be one line

`return strings.EqualFold(cdt.BaseDenom, cdt2.BaseDenom) && strings.EqualFold(cdt.Threshold, cdt2.Threshold)`

File: x/oracle/keeper/msg_server.go [hunk 4]
```diff
@@ -227,8 +227,23 @@ func (ms msgServer) GovAddDenoms(
 			return nil, err
 		}
 	}
-	// also update RewardBand key
-	plan.Keys = append(plan.Keys, string(types.KeyRewardBands))
+
+	// append new currency pair providers
+	for _, cpp := range msg.CurrencyPairProviders {
+		plan.Keys = ojoutils.AppendUniqueString(plan.Keys, string(types.KeyCurrencyPairProviders))
```
Comment (zarazan) [inline x/oracle/keeper/msg_server.go:233]: Is there a way to remove currency pair providers with this implementation?

**Review by zarazan — APPROVED** [review]

File: x/oracle/keeper/msg_server.go (hunk 4 above)
Comment (rbajollari) [reply x/oracle/keeper/msg_server.go:234]: [debd18c](https://github.com/ojo-network/ojo/pull/259/commits/debd18c3bb5b8fc06fbc1bc09538372acf59074c)

**Pull request merged**